horovod/common/stall_inspector.cc

```diff
@@ -0,0 +1,158 @@
+               "are waiting for remainder of ranks for more than "
+            << stall_warning_time.count() << " seconds. "
+            << "This may indicate that different ranks are trying to "
+               "submit different tensors or that only subset of ranks is "
+               "submitting tensors, which will cause deadlock. "
+            << std::endl
+            << "Stalled ranks:";
+    for (auto& kv : missing_ranks) {
+      message << std::endl << kv.first;
+      if (shutdown_ranks.find(kv.first) != shutdown_ranks.end()) {
+        message << "!";
+      }
+
+      message << ": [";
+      auto it = kv.second.begin();
+      message << *it;
+      int count = 0;
+      while (++it != kv.second.end()) {
+        message << ", " << *it;
+        if (++count == 5) {
+          message << " ...";
+          break;
+        }
+      }
+
+      message << "]";
+    }
+
+    if (should_shut_down) {
+      message
+          << std::endl
+          << "One or more rank (marked by \"!\") is stalled for longer than "
+          << stall_shutdown_time.count() << " seconds. Will shutdown.";
+      LOG(ERROR) << message.str();
+    } else {
+      LOG(WARNING) << message.str();
+    }
+  }
+
+  return should_shut_down;
+}
+
+// Invalidate cached tensors that have been pending for a long time.
+void StallInspector::InvalidateStalledCachedTensors(
+    CacheCoordinator& cache_coordinator, ResponseCache& response_cache) {
+  auto now = std::chrono::steady_clock::now();
+  std::chrono::seconds stall_warning_time(stall_warning_time_seconds);
+
+  for (auto& entry : cache_tensor_start) {
+    // If pending time for cached tensor exceeds stall_warning_time, mark entry
+    // for global removal from cache to trigger stall messaging.
+    if (now - entry.second > stall_warning_time) {
+      uint32_t cache_bit = response_cache.peek_cache_bit(entry.first);
+      cache_coordinator.record_invalid_bit(cache_bit);
+      cache_coordinator.set_uncached_in_queue(true);
+    }
+  }
+}
+
+// Record initial time cached tensor is encountered in queue.
+void StallInspector::RecordInitialTime(const std::string& tensor_name) {
+  if (perform_stall_check &&
+      cache_tensor_start.find(tensor_name) == cache_tensor_start.end()) {
+    cache_tensor_start[tensor_name] = std::chrono::steady_clock::now();
+  }
+}
+
+// Remove timing entry if uncached or marked invalid.
+void StallInspector::RemoveEntry(const std::string& tensor_name) {
+  if (perform_stall_check) {
+    cache_tensor_start.erase(tensor_name);
+  }
+}
+
+bool StallInspector::ShouldPerformCheck() {
+  return perform_stall_check &&
+         std::chrono::steady_clock::now() - last_stall_check >
+             std::chrono::seconds(stall_warning_time_seconds);
+}
+void StallInspector::UpdateCheckTime() {
+  last_stall_check = std::chrono::steady_clock::now();
+}
+
+void StallInspector::SetPerformStallCheck1(bool perform_stall_check) {
+  StallInspector::perform_stall_check = perform_stall_check;
```

Comment:
Can call argument `value` and simplify to: `perform_stall_check = value`. Similarly fix the other methods.